### PR TITLE
[Fix #2744] Increase amount of logged messages

### LINKED ISSUE #2744: More debugging statements in verbose mode
Currently, there are still some `try`/`catch` blocks that ignore raise exceptions. For example: https://github.com/facebook/osquery/blob/master/osquery/config/config.cpp#L387

There should instead transition over to these: https://github.com/facebook/osquery/blob/master/osquery/config/config.cpp#L249

### PR DESCRIPTION
**Pull request by vjon**
Here's my attempt at increasing the amount of logging throughout the code base. Notes on convention:

- Messages are logged immediately after the thrown error, with the error itself (i.e. "e.what()") included with the return Status as applicable
- In cases where the error causes program termination, LOG(ERROR) is used (this one's obvious)
- In cases where the error causes the function to return early, LOG(WARNING) is used
- In cases where the error doesn't cause the function to return early, but alters its behavior noticeably, VLOG(1) is used
- In cases similar to the above, but where the condition could happen multiple times in quick succession (e.g. try/catch nested inside a loop), VLOG(2) is used

Notably absent from the above scheme is the use of LOG(FATAL), LOG(INFO), and higher levels of VLOG(...)

If there are any recommendations for better use of the various google-glog macros, I'd be happy to discuss them :)

### CODE REVIEW

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by PoppySeedPlehzr** [comment]
This is super awesome. Thanks for powering through this!

**Comment by osqueryer** [comment]
:-1: The commit 4393e9c3cc7cb3ea16dc250fea52f04d1e7819b8 (Job results: [110](https://jenkins.osquery.io/job/osqueryPullRequestOSX/110/)) failed one or more tests (**macOS/OS X**).

**Comment by osqueryer** [comment]
:-1: The commit 4393e9c3cc7cb3ea16dc250fea52f04d1e7819b8 (Job results: [826](https://jenkins.osquery.io/job/osqueryPullRequestWindows/826/)) failed one or more tests (**Windows**).

File: osquery/tables/utility/file.cpp [hunk 1]
```diff
@@ -151,7 +151,8 @@ QueryData genFile(QueryContext& context) {
       for (; begin != end; ++begin) {
         genFileInfo(begin->path(), directory_string, "", results);
       }
-    } catch (const fs::filesystem_error& /* e */) {
+    } catch (const fs::filesystem_error& e) {
+      VLOG(2) << "Failed to generate info for file: " << e.what();
```
Comment (theopolis) [inline osquery/tables/utility/file.cpp]: nit, let's stick with `1` for all the verbose logging.

File: osquery/tables/system/posix/suid_bin.cpp [hunk 2]
```diff
@@ -113,6 +113,7 @@ void genSuidBinsFromPath(const std::string& path, QueryData& results) {
       try {
         ++it;
       } catch (fs::filesystem_error& e) {
+        LOG(WARNING) << "Failed recovery from binary read error: " << e.what();
```
Comment (theopolis) [inline osquery/tables/system/posix/suid_bin.cpp]: This is more likely a `"Cannot enumerate files in directory: " + subpath`

File: osquery/tables/system/linux/processes.cpp [hunk 3]
```diff
@@ -233,6 +233,7 @@ static inline SimpleProcStat getProcStat(const std::string& pid) {
     try {
       stat.start_time = TEXT(AS_LITERAL(BIGINT_LITERAL, details.at(19)) / 100);
     } catch (const boost::bad_lexical_cast& e) {
+      VLOG(1) << "Failed to set stat start time";
```
Comment (theopolis) [inline osquery/tables/system/linux/processes.cpp]: This doesn't provide much help, what PID is it referring to?

File: osquery/tables/system/linux/processes.cpp [hunk 4]
```diff
@@ -163,9 +163,9 @@ void genProcessMap(const std::string& pid, QueryData& results) {
     try {
       auto offset = std::stoll(fields[2], nullptr, 16);
       r["offset"] = (offset != 0) ? BIGINT(offset) : r["start"];
-
     } catch (const std::exception& e) {
       // Value was out of range or could not be interpreted as a hex long long.
+      VLOG(1) << "Error converting value: " << e.what();
```
Comment (theopolis) [inline osquery/tables/system/linux/processes.cpp]: This is missing some context, how do I know what to start investigating? Maybe add a reference to `offset` or `start` and the PID?

File: osquery/tables/system/linux/processes.cpp [hunk 5]
```diff
@@ -154,7 +154,7 @@ void genProcessMap(const std::string& pid, QueryData& results) {
         r["start"] = "0x" + addresses[0];
         r["end"] = "0x" + addresses[1];
       } else {
-        // Problem with the address format.
+        VLOG(1) << "Faulty address format";
```
Comment (theopolis) [inline osquery/tables/system/linux/processes.cpp]: This needs more context.

File: osquery/remote/serializers/json.cpp [hunk 6]
```diff
@@ -21,7 +21,8 @@ Status JSONSerializer::serialize(const pt::ptree& params,
   try {
     pt::write_json(output, params, false);
   } catch (const pt::json_parser::json_parser_error& e) {
-    return Status(1, std::string("JSON serialize error: ") + e.what());
+    LOG(WARNING) << "Error serializing JSON: " << e.what();
```
Comment (theopolis) [inline osquery/remote/serializers/json.cpp]: I'm worried about these callsites where the JSON error isn't very helpful until you know what is attempting to serialize the JSON. Perhaps the log should be done one frame above?

File: osquery/logger/plugins/filesystem.cpp [hunk 7]
```diff
@@ -104,6 +104,7 @@ Status FilesystemLoggerPlugin::logStringToFile(const std::string& s,
                            FLAGS_logger_mode,
                            true);
   } catch (const std::exception& e) {
+    LOG(WARNING) << "Error writing to file: " << e.what();
```
Comment (theopolis) [inline osquery/logger/plugins/filesystem.cpp]: What file was it? Perhaps this should be logged one frame above?

**Comment by theopolis** [comment]
This is really good work!

I'm cautious about the impact on verbosity vs increased effectiveness on debugging. Some general questions:
- Have these changes been tested in practice and shown to not overwhelm `--verbose=true` and `--verbose=false` modes.
- Does the extra verbosity make errors easier to find? This is a difficult question, I know, but there are several examples in the `osquery_tests` and `osquery_additional_tests` where the output may provide a gut-check, same with the python-based tests.
- Have we performed a quick review of the "call sites", 1 or 2 frames above each method? I know this is also a large ask, but there are some instances where we would now log in parsing / utility code, and the call site is choosing to log almost the same message.

I am pulling this PR in locally and giving it a spin. :)

**Comment by vjon** [comment]
@theopolis, thanks for your feedback. I've fixed some of the more straightforward reviews you made, and some more thoughts on your notes follow:

- Regarding verbose logging, I'm honestly unsure that my usage for osquery is anywhere near high-volume enough to properly stress test and see the worst-case performance. For what it's worth, my limited testing thus far seems to suggest it's fine. I'll continue to look for better ways to test this.
- Regarding surfacing errors, I'm going to give some more thought on how to optimize this. My first thought for starters would be to force the test suite to fail in multiple ways and see how useful the log output is.
- Regarding appropriate log site, this is going to be the area that requires the most attention. For my first pass, I made all the logging calls at consistent locations, just to make it easier to wrap ones head around the problem. There is definitely room for improvement here (as you pointed out in the case of the JSON (de)serialization example).
- Related to the log sites, on my next pass(es), I think it might be useful to select `VLOG` vs `LOG` on a more case-by-case basis. In particular, having all early function termination statements use `LOG` may be "over-stating" the importance of some functions, for which `VLOG` may actually be more appropriate.

On top of all this, I'm fairly certain that there's some places left which would benefit from increased logging, which I missed on my first pass. I'll probably end up doing a fairly close reading of most of the code, as I can't think of a great way to `grep` for potential logging areas.

File: osquery/remote/serializers/json.cpp (hunk 6 above)
Comment (vjon) [inline osquery/remote/serializers/json.cpp]: I think you might be right, as the only thing which can be provided for context within the serialize/deserialize functions are the string/pt::tree being parsed, which would probably not provide the necessary context (and clog up the logs as well).

**Comment by theopolis** [comment]
Solid, thank you so much for putting time into this!

I can cut a version from your branch and deploy to a few thousand hosts next week, then tally up the log lines for the stable vs. unstable versions.

**Comment by vjon** [comment]
That would be a great help, thanks! Looks like I have the rest of the week to put a few polishing touches on this commit.

**Comment by osqueryer** [comment]
:-1: The commit cf513f098a090f8f43986286afcf75698dc8ff78 (Job results: [3898](https://jenkins.osquery.io/job/osqueryPullRequestLinux/3898/)) failed one or more tests (**Linux**).

**Comment by vjon** [comment]
Okay, apart from the last outstanding issue re: JSON serializer log messages, this code is ready for testing. I've already found a number of places where further logging could be useful, but the code as is should provide a useful data point w.r.t. the verbosity/performance trade-off.

**Comment by vjon** [comment]
ping @theopolis

**Comment by theopolis** [comment]
Ah, sorry for my absence. Code reviews are high-pri for me tomorrow afternoon!

**Comment by jtizzle8369** [comment]
Dear valued potential customers:

Here at Honda we have been well known for over 20 years for providing the
best in reliabillity, comfort, and style. Over the years we have risen to
be one of the top auto industries here in Japan. But that isn't enough. We
want to be number one in the US. Now our twentieth anniversary for making
cars is here!!! This is the perfect opportunity for you and us here at
Honda to celebrate our 20 years of excellent service. We have been trying
to think of ideas to get more people to know about our cars. And with
technology and e-mail being the wave of the future, we want to jump on this
opportunity. So we have set up a rewards system to repay those who help us
spread the word about Honda.

Our marketing staff has designed a special program that traces this message
as it travels across the US. Anyone who forwards this e-mail, will
immediately have an account at their local Honda dealer opened in their
name. This account will initially be opened with a credit of $1,000 toward
any new or used vehicle at their participating dealership. For each person
you forward this e-mail to, the amount of $200 will be added to your
account. If the recipients of this e-mail forward it you will be rewarded
an additional $100 for each person it reaches and if they also forward it
your account continues to grow in $100 increments.

You can log onto our website at http://www.Honda.com to check the balance
of your account. If things go well and everyone participates you should see
your account grow quite quickly. Follow the on screen instructions to order
the specific make and model of Honda you want to buy with your account.

We hope that this is a rewarding experience for you and us. Our goal is to
reach over 1 million computers by the year 2000. I thank you for your time
and business,

Sincerely,

Kageyama Hironobu
Senior Honda Marketing Advisor

jdog83

File: osquery/core/flags.cpp [hunk 8]
```diff
@@ -44,6 +45,7 @@ int Flag::createAlias(const std::string& alias, const FlagDetail& flag) {
 Status Flag::getDefaultValue(const std::string& name, std::string& value) {
   flags::CommandLineFlagInfo info;
   if (!flags::GetCommandLineFlagInfo(name.c_str(), &info)) {
+    LOG(WARNING) << "Flags name not found.";
```
Comment (theopolis) [inline osquery/core/flags.cpp]: This should include the name of the missing flag.

File: osquery/core/flags.cpp [hunk 9]
```diff
@@ -95,7 +97,8 @@ Status Flag::updateValue(const std::string& name, const std::string& value) {
     flags::SetCommandLineOption(real_name.c_str(), value.c_str());
     return Status(0, "OK");
   }
-  return Status(1, "Flag not found");
+  LOG(WARNING) << "Flag not found.";
```
Comment (theopolis) [inline osquery/core/flags.cpp]: Same.

File: osquery/core/flags.cpp [hunk 10]
```diff
@@ -105,7 +108,7 @@ std::map<std::string, FlagInfo> Flag::flags() {
   std::map<std::string, FlagInfo> flags;
   for (const auto& flag : info) {
     if (instance().flags_.count(flag.name) == 0) {
-      // This flag info was not defined within osquery.
+      VLOG(1) << "Flag not defined within osquery.";
```
Comment (theopolis) [inline osquery/core/flags.cpp]: Same.

File: osquery/core/flags.cpp [hunk 11]
```diff
@@ -161,7 +164,7 @@ void Flag::printFlags(bool shell, bool external, bool cli) {
         continue;
       }
     } else {
-      // This flag was not defined as an osquery flag or flag alias.
+      VLOG(1) << "Flag or flag alias not defined within osquery.";
```
Comment (theopolis) [inline osquery/core/flags.cpp]: Same.

File: osquery/core/tables.cpp [hunk 12]
```diff
@@ -91,7 +92,9 @@ void TablePlugin::setContextFromRequest(const PluginRequest& request,
     std::stringstream input;
     input << request.at("context");
     pt::read_json(input, tree);
-  } catch (const pt::json_parser::json_parser_error& /* e */) {
+  } catch (const pt::json_parser::json_parser_error& e) {
+    LOG(WARNING) << "Error deserializing content from JSON: Tables Core ("
```
Comment (theopolis) [inline osquery/core/tables.cpp]: Why is "Tables Core" included here?

File: osquery/core/tables.cpp [hunk 13]
```diff
@@ -226,6 +231,7 @@ std::string columnDefinition(const PluginResponse& response, bool aliases) {
   std::map<std::string, ColumnType> column_types;
   for (const auto& column : response) {
     if (column.count("id") == 0) {
+      VLOG(1) << "Invalid column, skipping.";
```
Comment (theopolis) [inline osquery/core/tables.cpp]: I don't think this is offering much value.

File: osquery/database/database.cpp [hunk 14]
```diff
@@ -76,7 +77,8 @@ Status serializeRowJSON(const Row& r, std::string& json) {
   try {
     pt::write_json(output, tree, false);
   } catch (const pt::json_parser::json_parser_error& e) {
-    // The content could not be represented as JSON.
+    LOG(WARNING) << "Error serializing row as JSON: Database (" << e.what()
```
Comment (theopolis) [inline osquery/database/database.cpp]: I don't think you need the "Database" for these, each LOG will be decorated with the file name and line number.

File: osquery/database/query.cpp [hunk 15]
```diff
@@ -65,11 +65,11 @@ Status Query::addNewResults(const QueryData& current_qd,
   if (!isQueryNameInDatabase()) {
     // This is the first encounter of the scheduled query.
     fresh_results = true;
-    LOG(INFO) << "Storing initial results for new scheduled query: " << name_;
+    VLOG(1) << "Storing initial results for new scheduled query: " << name_;
```
Comment (theopolis) [inline osquery/database/query.cpp]: It's important to keep these as `LOG(INFO)`.

File: osquery/distributed/distributed.cpp [hunk 16]
```diff
@@ -108,7 +108,9 @@ Status Distributed::serializeResults(std::string& json) {
   try {
     pt::write_json(ss, results, false);
   } catch (const pt::ptree_error& e) {
-    return Status(1, "Error writing JSON: " + std::string(e.what()));
+    LOG(WARNING) << "Error serializing JSON: Distributed Subsystem ("
```
Comment (theopolis) [inline osquery/distributed/distributed.cpp]: Same with the "Distributed Subsystem"

File: osquery/distributed/distributed.cpp [hunk 17]
```diff
@@ -122,8 +124,8 @@ void Distributed::addResult(const DistributedQueryResult& result) {
 Status Distributed::runQueries() {
   while (getPendingQueryCount() > 0) {
     auto request = popRequest();
-    LOG(INFO) << "Executing distributed query: " << request.id << ": "
-              << request.query;
+    VLOG(1) << "Executing distributed query: " << request.id << ": "
```
Comment (theopolis) [inline osquery/distributed/distributed.cpp]: This is important to keep as `LOG(INFO)`.

File: osquery/events/kernel.cpp [hunk 18]
```diff
@@ -51,11 +51,13 @@ Status KernelEventPublisher::setUp() {
     queue_ = new CQueue(kKernelDevice, kKernelQueueSize);
   } catch (const CQueueException &e) {
     queue_ = nullptr;
+    LOG(WARNING) << "Error initializing CQueue object: " << e.what();
```
Comment (theopolis) [inline osquery/events/kernel.cpp]: The `CQueue` is a bit opaque to users, we should say something like "Error initializing kernel structures: "

File: osquery/logger/plugins/filesystem.cpp [hunk 19]
```diff
@@ -104,6 +104,8 @@ Status FilesystemLoggerPlugin::logStringToFile(const std::string& s,
                            FLAGS_logger_mode,
                            true);
   } catch (const std::exception& e) {
+    LOG(WARNING) << "Error writing to file " << (log_path_ / filename).string()
```
Comment (theopolis) [inline osquery/logger/plugins/filesystem.cpp]: nit, have we been enclosing exception text in parens?

**Comment by theopolis** [comment]
For:
> Related to the log sites, on my next pass(es), I think it might be useful to select VLOG vs LOG on a more case-by-case basis. In particular, having all early function termination statements use LOG may be "over-stating" the importance of some functions, for which VLOG may actually be more appropriate.

It seems like there's a large number of verbose log sites related to JSON parsing errors. Most are verbose logging now but generally we don't need to differentiate similar log text since GLog will decorate each log event with the filename and line number.

Also, we've been ignoring punctuation in log lines. I'm not sure why we started this way, but it's best to remain consistent.

**Comment by theopolis** [comment]
Some `LOG(INFO)` and `LOG(WARNING)` have been converted to VLOG. Some of these may be appropriate given the new increased logging, but do you mind isolating these changes into another follow up PR. I want to be clear that we're not introducing regressions for alarming/alerting folks have set up.

**Comment by vjon** [comment]
Okay, I've fixed the issues outlined and touched up things like the punctuation as well.

I'll get the second PR (for LOG(...) -> VLOG(1) conversions) ready in the near future.

**Comment by vjon** [comment]
Sorry for the absence, had some other business to take care of.

With respect to the last outstanding review, I still don't have a good solution for this. To summarize:
- Moving the logging statement up to the caller instead of inside the serialize/deserialize functions doesn't work in the long run. Since the rest of the code base has logging inside the relevant function itself, breaking from the convention for this case will undoubtedly result in new code being written that forgets to log errors returned from these functions.
- Using a field from the `pt::tree` doesn't work, as there isn't a field that is guaranteed to be in every PTree object created. Ideally, if the JSON functionality is ever re-factored to use JSON C++ libraries instead, there may be some information from such a `json` object that could be included in these error messages for context.

If anybody has any suggestions for potential approaches to resolve this issue, please let me know.

**Pull request closed without merging**